<a href="https://colab.research.google.com/github/virajvaidya/OptionsPricing/blob/main/BlackScholes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [36]:
#Basic dependencies
import numpy             as np
import pandas            as pd

In [ ]:
#For options pricing
!pip install py_vollib

In [ ]:
!pip install py_vollib_vectorized
import py_vollib_vectorized

In [40]:
from __future__                                  import division
from py_vollib.black                             import undiscounted_black
from py_vollib.black_scholes                     import black_scholes
from py_vollib.black_scholes                     import greeks
from py_vollib.black_scholes                     import black_scholes
from py_vollib.black_scholes.implied_volatility  import implied_volatility
from py_vollib_vectorized                        import api
from py_vollib_vectorized.api                    import get_all_greeks
from py_vollib_vectorized.api                    import price_dataframe

#Computing the Black-Scholes Option Price and the Implied Volatility

Stating the parameters:

**S** = Price of underlying asset

**K**        = Strike Price

**sigma**    = volatility (annualised)

**t**        = time to maturity in years (eg 200/365 implies 200 days to maturity)

**r** = risk free interest rate

**flag** = 'c' for call, 'p' for put

In [48]:
S         = 100                     
K         = 100
sigma     = 0.1 
t         = 250/365
r         = 0.1
flag      = 'c'

In [49]:
price = black_scholes(flag, S, K, t, r, sigma)        #Black Scholes Option Price  
iv = implied_volatility(price, S, K, t, r, flag)      #Black Scholes Implied Volatility

In [50]:
print (price,iv)

   Price
0  7.535     IV
0  0.1


**Interpretation of the above result:**

For a call option where

Underlying asset price = $100

Strike price = $100

annualised volatility = 10%

Days to maturity = 250 days

Risk-free Interest Rate = 10%

the Black-Scholes Option Price is $ 7.534

and

the Black-Scholes Implied Volatility is 10%

#Greeks

In [51]:
flag = ['c', 'p']
S = 100
K = 100
t = 250/365
r = .1
sigma = .1
get_all_greeks(flag, S, K, t, r, sigma, model='black_scholes', return_as='dict')

{'delta': [0.8075725925622557, -0.19242740743803943],
 'gamma': [0.033045366905071205, 0.03304536702386507],
 'rho': [0.5013229186263488, -0.13827102468310354],
 'theta': [-0.024592467737197587, 0.000994594877669952],
 'vega': [0.22573628568155657, 0.22573628568155601]}

**Interpretation of the above:**

The two numbers in the brackets provide the call and put Greeks respectively.

For example,

'delta': [0.8075725925622557, -0.19242740743803943]

Here the the delta for the call is 0.80758 and for the put is -0.192427

**However, the above process just for one option is cumbersome. We can calculate all the greeks for a dataframe as following:**

#Calculating greeks for a dataframe
(Randomly generated values for the variables)

In [156]:
df = pd.DataFrame()
df['Flag'] = ('c','c', 'c', 'c', 'c', 'c', 'c')
df['S'] = (95, 100, 53, 56,67,98,123)
df['K'] = (100, 100, 56, 57,70,100,120)
df['T'] = (0.2, 0.1, 0.15,0.12, 0.08, 0.09, 0.14)
df['R'] = (0.2, 0.3, 0.1, 0.1, 0.1, 0.1, 0.05)
df['IV'] = (0.2, 0.4, 0.12, 0.34, 0.21, 0.1, 0.1)
price_dataframe(df, flag_col='Flag', underlying_price_col='S', strike_col='K', annualized_tte_col='T', riskfree_rate_col='R', sigma_col='IV', model='black_scholes', inplace=False)

Price     delta     gamma     theta       rho      vega
0  2.895588  0.467506  0.046795 -0.045900  0.083035  0.168926
1  6.584992  0.618070  0.030148 -0.111946  0.055222  0.120588
2  0.270435  0.200816  0.113935 -0.009119  0.015562  0.057496
3  2.480319  0.504188  0.060482 -0.037260  0.030905  0.077386
4  0.680141  0.283299  0.085066 -0.028133  0.014641  0.064130
5  0.708719  0.360013  0.127252 -0.026223  0.031117  0.109918
6  4.337847  0.806680  0.059593 -0.025397  0.132828  0.125875

In [157]:
df = pd.DataFrame()
df['Flag'] = ('p','p', 'p', 'p', 'p', 'p', 'p')
df['S'] = (95, 100, 53, 56,67,98,123)
df['K'] = (100, 100, 56, 57,70,100,120)
df['T'] = (0.2, 0.1, 0.15,0.12, 0.08, 0.09, 0.14)
df['R'] = (0.2, 0.3, 0.1, 0.1, 0.1, 0.1, 0.05)
df['IV'] = (0.2, 0.4, 0.12, 0.34, 0.21, 0.1, 0.1)
price_dataframe(df, flag_col='Flag', underlying_price_col='S', strike_col='K', annualized_tte_col='T', riskfree_rate_col='R', sigma_col='IV', model='black_scholes', inplace=False)

Price     delta     gamma     theta       rho      vega
0  3.974532 -0.532494  0.046795  0.006761 -0.109123  0.168926
1  3.629545 -0.381930  0.030148 -0.032151 -0.041823  0.120588
2  2.436703 -0.799184  0.113935  0.005997 -0.067187  0.057496
3  2.800407 -0.495812  0.060482 -0.021828 -0.036679  0.077386
4  3.122375 -0.716701  0.085066 -0.009106 -0.040913  0.064130
5  1.812757 -0.639987  0.127252  0.000932 -0.058077  0.109918
6  0.500780 -0.193320  0.059593 -0.009072 -0.034001  0.125875